src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
-
-tyVarBndrName :: TyVarBndr -> Name
-tyVarBndrName (PlainTV n) = n
-tyVarBndrName (KindedTV n _) = n
-
-tyVarBndrKind :: TyVarBndr -> Maybe Type
-tyVarBndrKind (PlainTV _) = Nothing
-tyVarBndrKind (KindedTV _ k) = Just k
-
--- | Generates a function type from the corresponding GADT type constructor
--- @x :: Member (Effect e) r => a -> b -> c -> Sem r r@.
-genType :: Con -> Q (Type, Maybe Name, Maybe Type)
-genType (ForallC tyVarBindings conCtx con) = do
-  (t, mn, _) <- genType con
-  let k = do n <- mn
-             z <- find ((== n) . tyVarBndrName) tyVarBindings
-             tyVarBndrKind z
-      free = everything mappend freeVars t
-  pure ( ForallT (filter (flip elem free . tyVarBndrName) tyVarBindings) conCtx t
-       , mn
-       , k
-       )
-genType (GadtC   _ tArgs' (eff `AppT` m `AppT` tRet)) = do
-  r <- newName "r"
-  let
-    tArgs            = fmap snd tArgs'
-    memberConstraint = ConT ''Member `AppT` eff `AppT` VarT r
-    resultType       = ConT ''Sem `AppT` VarT r `AppT` tRet
-
-    replaceMType t | t == m = ConT ''Sem `AppT` VarT r
-                   | otherwise = t
-    ts = everywhere (mkT replaceMType) tArgs
-    tn = case tRet of
-           VarT n -> Just n
-           _ -> Nothing
-
-  pure
-    . (, tn, Nothing)
-    .  ForallT [PlainTV r] [memberConstraint]
-    .  foldArrows
-    $  ts
-    ++ [resultType]
--- TODO: Although this should never happen, we obviously need a better error message below.
-genType _       = fail "genSig expects a GADT constructor"
-
--- | Turn all (KindedTV tv StarT) into (PlainTV tv) in the given type
--- This can prevent the need for KindSignatures
-simplifyBndrs :: Maybe Type -> Type -> Type
-simplifyBndrs star = everywhere (mkT $ simplifyBndr star)
-
--- | Turn TvVarBndrs of the form (KindedTV tv StarT) into (PlainTV tv)
--- This can prevent the need for KindSignatures
-simplifyBndr :: Maybe Type -> TyVarBndr -> TyVarBndr
-simplifyBndr (Just star) (KindedTV tv k) | star == k = PlainTV tv
-simplifyBndr _ (KindedTV tv StarT) = PlainTV tv
-simplifyBndr _ bndr = bndr
-
--- | Generates a type signature of the form
--- @x :: Member (Effect e) r => a -> b -> c -> Sem r r@.
-genSig :: Con -> Q Dec
-genSig con = do
-  let
-    getConName (ForallC _ _ c) = getConName c
-    getConName (GadtC [n] _ _) = pure n
-    getConName c = fail $ "failed to get GADT name from " ++ show c
-  conName <- getConName con
-  (t, _, k) <- genType con
-  pure $ SigD (getDeclName conName) $ simplifyBndrs k t
 
+------------------------------------------------------------------------------
 -- | Folds a list of 'Type's into a right-associative arrow 'Type'.
 foldArrows :: [Type] -> Type
-foldArrows = foldr1 (AppT . AppT ArrowT)
+foldArrows = foldr1 $ AppT . AppT ArrowT
+
+------------------------------------------------------------------------------
+-- | Extracts name from type variable (possibly nested in signature and/or
+-- some context), returns 'Nothing' otherwise.
+tvName' :: Type -> Maybe Name
```

Comment:
why the prime?

Reply:
`tvName` from `th-abstraction` works on `TyVarBndrs` - I wanted to differentiate them somehow, but `tvName` is the most sensible name for both...

Reply:
Changed to `tVarName`.